sdks/python/apache_beam/io/gcp/pubsub.py

```diff
@@ -314,6 +351,26 @@ def expand(self, pcoll):
     return pcoll | WriteToPubSub(self.topic)
 
 
+class _AddMetricsAndMap(DoFn):
+  def __init__(self, fn, project, topic=None):
+    self.project = project
+    self.topic = topic
+    self.fn = fn
+    self.reported_lineage = False
+
+  def setup(self):
+    self.reported_lineage = False
+
+  def process(self, element):
+    self.report_lineage_once()
+    yield self.fn(element)
+
+  def report_lineage_once(self):
```

Comment:
Do we need to ensure read/write on report_lineage is atomic? In other words, is it possible that multiple threads read and write reported_lineage at the same time leading to a race condition?

Reply:
It does not matter because Lineage is backed by a String Set, report once and multiple times has the same result (idempotent). Here I use a local variable to reduce the overhead a little bit (do not do set add operation on every element)

Reply:
This pretty much mirrors Java SDK change of #32068 (BigtableIO), #32090 (FileIO), #32381 (PubSubIO, final version after feedbacks)

Reply:
ok, i see.